# Add dbGaP and CDS Submission Status to WUSTL's File Metadata
## Date: 2022-02-22
## Author: Jeffrey Grover
**Purpose:** Previously, we had made copies of files and added this as tags. That is bad, and the multiple copies will make maintaining the metadata in the datapool even more of a disaster than it is. With multiple copies we don't know which one is the "main" copy and should get metadata applied. So, I'm going to add the dbGaP and CDS submission status of everything as more metadata fields to the "main" files.

### Load libraries

In [9]:
library(tidyverse)
library(sevenbridges)

### Load the metadata
They had some files which hadn't already been analyzed, and those were in a separate folder on the CGC. I added the `dbgap_accession` and `cds_status` as metadata fields to those separate, and put them in the `Data_Submission` folder under `WUSTL`.

There are also a bunch of files which were previously existing in the WUSTL folders. There may be a second copy in the dbGaP folder. Therefore, I don't know which ones should be the "main" copy and get the dbGaP and CDS submission metadata. To fix this, I can compare the lists of files.

In [4]:
existing_wustl_files <- read_csv('2022-02-22_wustl_datapool_existing_rnaseq_wes_manifest.csv') %>%
    mutate(name = basename(name))

head(existing_wustl_files)

Rows: 434 Columns: 66

── Column specification ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
chr (19): id, name, project, passage, sample_id, sample_type, library_prep_k...
dbl  (2): size, paired_end
lgl (45): trimmed_read_count, p1_1, Is FFPE, prop_ribosomal_bases, Restricti...


ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.



id                       name                                    size       
1 5d8914ace4b06d0cb492458b WUPDTC-002-B01.pe_2.fastq.paired.fq     14208130600
2 5d891c1fe4b065d300c6e31c WUPDTC-002-T01-P12.pe_1.fastq.paired.fq 18071815881
3 5d891c1fe4b065d300c6e31b WUPDTC-002-T01-P12.pe_2.fastq.paired.fq 18071815881
4 5d8914bee4b06d0cb4924592 WUPDTC-002-T01.pe_1.fastq.paired.fq     13816487982
5 5d8914bee4b06d0cb4924590 WUPDTC-002-T01.pe_2.fastq.paired.fq     13816487982
6 5d891395e4b065d300c6a1fc WUPDTC-003-B01.pe_1.fastq.paired.fq     13787659314
  project                trimmed_read_count p1_1 Is FFPE passage
1 pdxnet/pdxnet-datapool NA                 NA   NA      NA     
2 pdxnet/pdxnet-datapool NA                 NA   NA      P12    
3 pdxnet/pdxnet-datapool NA                 NA   NA      P12    
4 pdxnet/pdxnet-datapool NA                 NA   NA      NA     
5 pdxnet/pdxnet-datapool NA                 NA   NA      NA     
6 pdxnet/pdxnet-datapool NA                 NA   NA      NA     
  sample_id          prop_ribosomal_bases ⋯ dbgap_upload model_id paired_end
1 WUPDTC-002-B01     NA                   ⋯ NA           NA       2         
2 WUPDTC-002-T01-P12 NA                   ⋯ NA           NA       1         
3 WUPDTC-002-T01-P12 NA                   ⋯ NA           NA       2         
4 WUPDTC-002-T01     NA                   ⋯ NA           NA       1         
5 WUPDTC-002-T01     NA                   ⋯ NA           NA       2         
6 WUPDTC-003-B01     NA                   ⋯ NA           NA       1         
  __inherit__ update_manifest species tumor_id investigation prop_20x_cov
1 NA          NA              NA      NA       NA            NA          
2 NA          NA              NA      NA       NA            NA          
3 NA          NA              NA      NA       NA            NA          
4 NA          NA              NA      NA       NA            NA          
5 NA          NA              NA      NA       NA            NA          
6 NA          NA              NA      NA       NA            NA          
  experimental_strategy
1 WES                  
2 WES                  
3 WES                  
4 WES                  
5 WES                  
6 WES

In [6]:
cds_wustl_files <- read_csv('2022-02-22_wustl_datapool_cds_rnaseq_wes_manifest.csv') %>%
    mutate(name = basename(name))

head(cds_wustl_files)

Rows: 850 Columns: 66

── Column specification ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
chr (10): id, name, project, sample_type, case_id, patient_id, contributor, ...
dbl  (1): size
lgl (55): trimmed_read_count, p1_1, Is FFPE, passage, sample_id, prop_riboso...


ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.



id                       name                            size      
1 6053a7661f537262530d81d0 WU-L034-T1.WES.r1.fastq.gz      2643723290
2 6053a7661f537262530d81c2 WU-L034-T1.WES.r2.fastq.gz      2746938346
3 6053a7661f537262530d81c3 WU-L035-N1.WES.r1.fastq.gz      1876415309
4 6053a7661f537262530d81f9 WU-L035-N1.WES.r2.fastq.gz      1886646726
5 6053a7661f537262530d827e WU-L035-T1-X1P0.WES.r1.fastq.gz  837482288
6 6053a7661f537262530d824a WU-L035-T1-X1P0.WES.r2.fastq.gz  859970829
  project                trimmed_read_count p1_1 Is FFPE passage sample_id
1 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
2 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
3 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
4 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
5 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
6 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
  prop_ribosomal_bases ⋯ dbgap_upload model_id   paired_end __inherit__
1 NA                   ⋯ NA           NA         NA         NA         
2 NA                   ⋯ NA           NA         NA         NA         
3 NA                   ⋯ NA           NA         NA         NA         
4 NA                   ⋯ NA           NA         NA         NA         
5 NA                   ⋯ NA           WU-L035-T1 NA         NA         
6 NA                   ⋯ NA           WU-L035-T1 NA         NA         
  update_manifest species tumor_id   investigation prop_20x_cov
1 NA              NA      NA         NA            NA          
2 NA              NA      NA         NA            NA          
3 NA              NA      NA         NA            NA          
4 NA              NA      NA         NA            NA          
5 NA              NA      WU-L035-T1 NA            NA          
6 NA              NA      WU-L035-T1 NA            NA          
  experimental_strategy
1 WES                  
2 WES                  
3 WES                  
4 WES                  
5 WES                  
6 WES

### Find files missing from the existing files list

In [8]:
# Files not already in the datapool have not been analyzed
# Create a separate list

new_wustl_files <- cds_wustl_files %>% filter(!(name %in% existing_wustl_files$name))

nrow(new_wustl_files)
head(new_wustl_files)

[1] 478

id                       name                            size      
1 6053a7661f537262530d81d0 WU-L034-T1.WES.r1.fastq.gz      2643723290
2 6053a7661f537262530d81c2 WU-L034-T1.WES.r2.fastq.gz      2746938346
3 6053a7661f537262530d81c3 WU-L035-N1.WES.r1.fastq.gz      1876415309
4 6053a7661f537262530d81f9 WU-L035-N1.WES.r2.fastq.gz      1886646726
5 6053a7661f537262530d827e WU-L035-T1-X1P0.WES.r1.fastq.gz  837482288
6 6053a7661f537262530d824a WU-L035-T1-X1P0.WES.r2.fastq.gz  859970829
  project                trimmed_read_count p1_1 Is FFPE passage sample_id
1 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
2 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
3 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
4 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
5 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
6 pdxnet/pdxnet-datapool NA                 NA   NA      NA      NA       
  prop_ribosomal_bases ⋯ dbgap_upload model_id   paired_end __inherit__
1 NA                   ⋯ NA           NA         NA         NA         
2 NA                   ⋯ NA           NA         NA         NA         
3 NA                   ⋯ NA           NA         NA         NA         
4 NA                   ⋯ NA           NA         NA         NA         
5 NA                   ⋯ NA           WU-L035-T1 NA         NA         
6 NA                   ⋯ NA           WU-L035-T1 NA         NA         
  update_manifest species tumor_id   investigation prop_20x_cov
1 NA              NA      NA         NA            NA          
2 NA              NA      NA         NA            NA          
3 NA              NA      NA         NA            NA          
4 NA              NA      NA         NA            NA          
5 NA              NA      WU-L035-T1 NA            NA          
6 NA              NA      WU-L035-T1 NA            NA          
  experimental_strategy
1 WES                  
2 WES                  
3 WES                  
4 WES                  
5 WES                  
6 WES

### Move Files Which Haven't Been Analyzed Yet
To avoid chaos when it comes to file organization.

In [10]:
# Connect to the platform

sbg_auth <- Auth(from = 'file', profile_name = 'cgc')
sbg_proj <- sbg_auth$project(id = 'pdxnet/pdxnet-datapool')

Authenticating with user configuration file: ~/.sevenbridges/credentials

Authenticating with user profile: cgc



In [27]:
# Get the folder object for the analysis_incomplete folder

analysis_incomplete_folder <- sbg_proj$get_root_folder()$list_folder_contents(complete = TRUE)[[5]]$list_folder_contents(complete = TRUE)[[6]]$list_folder_contents(complete = TRUE)[[1]]

analysis_incomplete_folder
analysis_incomplete_folder$id

== Files ==
id : 62153e7819807d361ecf5e70
name : analysis_incomplete
project : pdxnet/pdxnet-datapool
parent : 620d1ca68b04ac66e6c00d5b
type : folder

[1] "62153e7819807d361ecf5e70"

In [30]:
# Test folder operations with the first file

# Test setting metadata
sbg_proj$file(id = new_wustl_files$id[1])$set_meta(cds_status = 'uploaded', dbgap_accession = 'phs002432.v1.p1')

# Test moving file
sbg_proj$file(id = new_wustl_files$id[1])$move_to_folder(analysis_incomplete_folder$id)

# Get file info
sbg_proj$file(id = new_wustl_files$id[1])
sbg_proj$file(id = new_wustl_files$id[1])$get_parent_folder()

$contributor
[1] "WUSTL"

$patient_id
[1] "WU-L034"

$case_id
[1] "WU-L034"

$cds_status
[1] "uploaded"

$experimental_strategy
[1] "WES"

$sample_type
[1] "Human_tumor"

$dbgap_accession
[1] "phs002432.v1.p1"

attr(,"response")
Response [https://cgc-api.sbgenomics.com/v2/files/6053a7661f537262530d81d0/metadata]
  Date: 2022-02-22 21:22
  Status: 200
  Content-Type: application/json
  Size: 184 B

== Files ==
id : 6053a7661f537262530d81d0
name : WU-L034-T1.WES.r1.fastq.gz
size : 2643723290
project : pdxnet/pdxnet-datapool
created_on : 2021-03-18T19:17:58Z
modified_on : 2022-02-22T21:22:43Z
storage:
  type : PLATFORM
  hosted_on_locations : 1 items
     - aws:us-east-1
tags:
   - phs002305.v1.p1
   - WES
   - 2020_10
   - WUSTL
   - CDS_DONE
metadata:
  contributor : WUSTL
  patient_id : WU-L034
  case_id : WU-L034
  cds_status : uploaded
  experimental_strategy : WES
  sample_type : Human_tumor
  dbgap_accession : phs002432.v1.p1
parent : 62153e7819807d361ecf5e70
type : file

== Files ==
id : 6053a7661f537262530d81d0
name : WU-L034-T1.WES.r1.fastq.gz
size : 2643723290
project : pdxnet/pdxnet-datapool
created_on : 2021-03-18T19:17:58Z
modified_on : 2022-02-22T21:22:43Z
storage:
  type : PLATFORM
  hosted_on_locations : 1 items
     - aws:us-east-1
tags:
   - phs002305.v1.p1
   - WES
   - 2020_10
   - WUSTL
   - CDS_DONE
metadata:
  contributor : WUSTL
  patient_id : WU-L034
  case_id : WU-L034
  cds_status : uploaded
  experimental_strategy : WES
  sample_type : Human_tumor
  dbgap_accession : phs002432.v1.p1
parent : 62153e7819807d361ecf5e70
type : file

== Files ==
id : 62153e7819807d361ecf5e70
name : analysis_incomplete
project : pdxnet/pdxnet-datapool
created_on : 2022-02-22T19:50:16Z
modified_on : 2022-02-22T19:50:16Z
parent : 620d1ca68b04ac66e6c00d5b
type : folder

In [32]:
# Loop to move the files which haven't been analyzed yet and set metadata fields

for (i in seq_along(new_wustl_files$id)) {
    
    # Pause for 5 mins every 200 iterations to avoid API call limits
    if (i %% 200 == 0) {
        Sys.sleep(300)
    }
    
    # Set metadata and move files
    sbg_proj$file(id = new_wustl_files$id[i])$set_meta(cds_status = 'uploaded', dbgap_accession = 'phs002432.v1.p1')
    sbg_proj$file(id = new_wustl_files$id[i])$move_to_folder(analysis_incomplete_folder$id)
}

All files which were uploaded to the datapool prior to cds and dbgap submissions are now in a separate folder to indicate that they haven't been analyzed.

### Add metadata to all cds-submitted files which already were on the CGC

In [36]:
# Filter the manifest for files which were submitted to CDS

existing_files_cds_uploaded <- existing_wustl_files %>% filter(name %in% cds_wustl_files$name)

nrow(existing_files_cds_uploaded)
head(existing_files_cds_uploaded)

[1] 372

id                       name                            size      
1 5d60005de4b09cae7680875a WU-0056-T1-X1P2.WES.r1.fastq.gz 2126007246
2 5d60005de4b09cae76808740 WU-0056-T1-X1P2.WES.r2.fastq.gz 2109115670
3 5d60005de4b09cae76808753 WU-0059-T1-X1P1.WES.r1.fastq.gz 2534091484
4 5d60005de4b09cae76808755 WU-0059-T1-X1P1.WES.r2.fastq.gz 2531199502
5 5f4efaf6e4b0bf4a931cc751 WU-0059-T1-X1P2.WES.r1.fastq.gz 3176500221
6 5f4efaf6e4b0bf4a931cc741 WU-0059-T1-X1P2.WES.r2.fastq.gz 3259841516
  project                trimmed_read_count p1_1 Is FFPE passage
1 pdxnet/pdxnet-datapool NA                 NA   NA      P2     
2 pdxnet/pdxnet-datapool NA                 NA   NA      P2     
3 pdxnet/pdxnet-datapool NA                 NA   NA      P1     
4 pdxnet/pdxnet-datapool NA                 NA   NA      P1     
5 pdxnet/pdxnet-datapool NA                 NA   NA      NA     
6 pdxnet/pdxnet-datapool NA                 NA   NA      NA     
  sample_id       prop_ribosomal_bases ⋯ dbgap_upload model_id paired_end
1 WU-0056-T1-X1P2 NA                   ⋯ NA           NA        1        
2 WU-0056-T1-X1P2 NA                   ⋯ NA           NA        2        
3 WU-0059-T1-X1P1 NA                   ⋯ NA           NA        1        
4 WU-0059-T1-X1P1 NA                   ⋯ NA           NA        2        
5 NA              NA                   ⋯ NA           NA       NA        
6 NA              NA                   ⋯ NA           NA       NA        
  __inherit__ update_manifest species tumor_id investigation prop_20x_cov
1 NA          NA              NA      NA       NA            NA          
2 NA          NA              NA      NA       NA            NA          
3 NA          NA              NA      NA       NA            NA          
4 NA          NA              NA      NA       NA            NA          
5 NA          NA              NA      NA       NA            NA          
6 NA          NA              NA      NA       NA            NA          
  experimental_strategy
1 WES                  
2 WES                  
3 WES                  
4 WES                  
5 WES                  
6 WES

In [37]:
# Add the metadata to the files in-place

for (i in seq_along(existing_files_cds_uploaded$id)) {
    
    # Pause for 5 mins every 200 iterations to avoid API call limits
    if (i %% 200 == 0) {
        Sys.sleep(300)
    }
    
    # Set metadata and move files
    sbg_proj$file(id = existing_files_cds_uploaded$id[i])$set_meta(cds_status = 'uploaded', dbgap_accession = 'phs002432.v1.p1')
}

The copy of those files still in the cds upload folder on the CGC can now be removed through the GUI. There are 186 RNAseq and 186 WES fastq files (total = 372), which all match the files that were elsewhere on the CGC, as shown above.